core/eolearn/core/eodata_io.py

```diff
@@ -382,4 +368,74 @@ def _decode(self, file, path):
                 data = data.set_geometry("geometry")
 
             return data
-        raise ValueError(f"Unsupported data type for file {path}.")
+
+        raise ValueError(f"Path {self.path} leads to a non-GPKG file, which is not supported.")
+
+    def _write_to_file(self, data: gpd.GeoDataFrame, file: Union[BinaryIO, gzip.GzipFile]) -> None:
+        layer = fs.path.basename(self.path)
+        try:
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    "ignore",
+                    message="You are attempting to write an empty DataFrame to file*",
+                    category=UserWarning,
+                )
+                return data.to_file(file, driver="GPKG", encoding="utf-8", layer=layer, index=False)
+        except ValueError as err:
+            # This workaround is only required for geopandas<0.11.0 and will be removed in the future.
+            if data.empty:
+                schema = infer_schema(data)
+                return data.to_file(file, driver="GPKG", encoding="utf-8", layer=layer, schema=schema)
+            raise err
+
+
+class FeatureIOjson(FeatureIO[_T]):
+    """FeatureIO object specialized for JSON-like objects."""
+
+    file_format = MimeType.JSON
+
+    def _read_from_file(self, file: Union[BinaryIO, gzip.GzipFile], _: str) -> _T:
+        return json.load(file)
+
+    def _write_to_file(self, data: _T, file: Union[BinaryIO, gzip.GzipFile]) -> None:
+        try:
+            json_data = json.dumps(data, indent=2, default=jsonify_timestamp)
+        except TypeError as exception:
+            raise TypeError(
+                f"Failed to serialize when saving JSON file to {self.path}. Make sure that this feature type "
+                "contains only JSON serializable Python types before attempting to serialize it."
+            ) from exception
+
+        file.write(json_data.encode())
+
+
+class FeatureIObbox(FeatureIO[BBox]):
+    """FeatureIO object specialized for BBox objects."""
+
+    file_format = MimeType.GEOJSON
+
+    def _read_from_file(self, file: Union[BinaryIO, gzip.GzipFile], _: str) -> BBox:
+        json_data = json.load(file)
+        return Geometry.from_geojson(json_data).bbox
+
+    def _write_to_file(self, data: BBox, file: Union[BinaryIO, gzip.GzipFile]) -> None:
+        json_data = json.dumps(data.geojson, indent=2)
+        file.write(json_data.encode())
+
+
+def jsonify_timestamp(param: object) -> str:
+    """Adds the option to serialize datetime.date objects via isoformat."""
+    if isinstance(param, datetime.date):
+        return param.isoformat()
+    raise TypeError(f"Object of type {type(param)} is not yet supported in jsonify utility function")
+
+
+def create_feature_io(ftype: FeatureType, path: str, filesystem: FS) -> FeatureIO:
```

Comment:
Maybe `jsonify_timestamp` and `create_feature_io` could start with an underscore so that we know they are private for this module?

Reply:
good call, undrescore added